python/code.qmd

```diff
@@ -0,0 +1,406 @@
+fundamental ways that are not always obvious.
+
+The way you avoid magic is by writing code that is simple, predictable,
+and consistent. This means that writing code in a more [functional
+programming](https://en.wikipedia.org/wiki/Functional_programming) style
+tends to produce more maintainable and more predictable code. Strategies
+for writing more functional code include:
+
+-   Using clear inputs and one type of output for functions and methods.
+    There is more explanation on this in the [type hints](#type-hints)
+    section below.
+-   Avoiding [side effects](#side-effects) by only outputting the result of the function
+    or method. This includes not modifying the input data or variables
+    by treating them as
+    [immutable](https://en.wikipedia.org/wiki/Immutable_object) (not
+    changeable).
+-   Minimizing internal conditional statements that change the action or
+    behaviour of the code.
+-   Preferring functionals like `map()`, `filter()`, and `reduce()` over
+    loops or list comprehensions, as they tend to be more predictable
+    and easier to read and write.
+
+::: callout-note
+Sometimes it's impossible to write in a purely functional style. For
+example, when you need to write to a file or database, this action is
+inherently a side effect. And that's why the most successful programming
+languages, like Python, are those that are flexible in the type of style
+you can use to write code. However, Python's functional programming
+capabilities are not as strong as those of some other languages, which
+can sometimes make it quite difficult to write functional code.
+:::
+
+A major benefit to writing more functional code is that it is much
+easier to create unit [tests](tests.qmd) for it and it is easier for the
+user to understand what the code does.
+
+### Type hints {#type-hints}
+
+For keeping the input and output consistent and clear, it helps to write
+[type hints](https://docs.python.org/3/library/typing.html). What are
+type hints in Python? They are a way to formally declare the type of
+object that is the input or output by a function or method. For example, for
+the function `check_heart_rate()`, you might write:
+
+``` {.python filename="src/health_data_checks/check_heart_rate.py"}
+def check_heart_rate(data: dict[str, Any]) -> bool:
+    """Check if the heart rate is within a normal range."""
+    # Code ...
+    return boolean_output
+```
+
+Where `data: dict[str, Any]` tells Python that the input to the function
+must be a [`dict`](https://realpython.com/python-dicts/) object with
+`str` or `Any` as internal data types and the `-> bool` tells Python
+that the output of the function has to be a boolean value (`True` or
+`False`). So if you look at the documentation for the function or
+method, you will be able to read what object types the code needs and
+what it outputs.
+
+::: callout-note
+Since you're using mypy when using a project made from
```

Comment:
Should this be re-phrased to say something like 'If you have based your project on our template, then you will be using mypy which requires type hints'?